### PR TITLE
[ENG-1911] Allow buying MintOffer

### PR DESCRIPTION
**Pull request by maximgeerinck**
# Description

Allows buying offers that are MintOffers. Basically lazy sale items.

### CODE REVIEW

File: src/offer/MintOffer.ts [hunk 1]
```diff
@@ -205,8 +212,78 @@ export class MintOffer extends BasicOffer {
...
+      this.offer.contract.contractAddress,
+      0,
+      this._offer.user?.ethAddress
+    );
+
+    const isNativeCurrency = this._chain.isNativeCurrency(
+      this._offer.price.currency
+    );
+
+    const message = {
```
Comment (sneljo1) [inline src/offer/MintOffer.ts:236]: Can we get some typings for this?

File: src/offer/MintOffer.ts [hunk 2]
```diff
@@ -205,8 +212,78 @@ export class MintOffer extends BasicOffer {
...
+    );
+
+    return new EvmTransaction(buyTx);
+  }
+
+  /**
+   * Singleton to get the corresponding lazy mint contract
+   * @returns ethers.Contract
+   */
+  async getContract(): Promise<ethers.Contract> {
```
Comment (sneljo1) [inline src/offer/MintOffer.ts:271]: Most of the functions in contracts actually already only go to the DB once.

File: src/refinable/Contracts.ts [hunk 3]
```diff
@@ -178,6 +181,23 @@ export class Contracts {
     return code !== "0x0";
   }
 
+  async findContract({
```
Comment (sneljo1) [inline src/refinable/Contracts.ts:184]: Could've also just used `getRefinableContract` and made types optional. If you decide to keep this, would also make use of the `this.cacheContract` and do the lazy loading inhere just like the other methods do

File: src/refinable/RefinableBaseClient.ts [hunk 4]
```diff
@@ -134,10 +130,7 @@ export abstract class RefinableBaseClient<O extends object = {}> {
     ) as GetUserOfferItemsQuery["user"]["itemsOnOffer"];
   }
 
-  async getOffer<O extends Offer = Offer>(
-    id: string,
-    storeId?: string
-  ): Promise<O> {
+  async getOffer(id: string, storeId?: string) {
```
Comment (sneljo1) [inline src/refinable/RefinableBaseClient.ts:133]: Can we keep the generic? Otherwise we can't automatically determine the offer type and methods on it

**Review by sneljo1 — CHANGES_REQUESTED** [review]
Looks good, couple of typing remarks. Can you also remove `yarn-error.log` and add to gitignore? And maybe add an e2e test for the claim?

File: src/offer/MintOffer.ts [hunk 5]
```diff
@@ -25,11 +31,12 @@ interface BuyParams {
 
 export class MintOffer extends BasicOffer {
   _chain: Chain;
+  _contract: ethers.Contract;
```
Comment (sneljo1) [inline src/offer/MintOffer.ts:34]: Should we make these private? Is that the goals of these as they are prefixed with _?

**Review by sneljo1 — CHANGES_REQUESTED** [review]

File: src/refinable/Contracts.ts (hunk 3 above)
Comment (maximgeerinck) [reply src/refinable/Contracts.ts:184]: would rather work with `findContract` than `getRefinableContract` though. getRefinableContract feels like it's implying that it's one of our tokens

File: src/offer/MintOffer.ts (hunk 5 above)
Comment (maximgeerinck) [reply src/offer/MintOffer.ts]: yep

File: src/refinable/Contracts.ts (hunk 3 above)
Comment (sneljo1) [reply src/refinable/Contracts.ts:184]: I'm fine either way, its indeed bad naming, also on the side of the api. But it does the same thing essentially.

File: src/offer/MintOffer.ts (hunk 2 above)
Comment (maximgeerinck) [reply src/offer/MintOffer.ts:268]: ? not sure what you mean?

File: src/refinable/RefinableBaseClient.ts [hunk 6]
```diff
@@ -21,33 +22,38 @@ import {
...
-}
-
 export abstract class RefinableBaseClient<O extends object = {}> {
   protected _apiClient?: GraphQLClient;
   protected _options: Options<O>;
   protected _apiKey: string;
-  protected accountAddress: string;
+  protected _accountAddress: string;
   protected account: Account;
+  public contracts: Contracts;
```
Comment (sneljo1) [inline src/refinable/RefinableBaseClient.ts:48]: Adding this type is not 100% correct, its misleading now because its only instantiated in the EvmClient

**Review by sneljo1 — COMMENTED** [review]
I also left a general comment about the yarn error log and the e2e tests I was promised :D

File: src/offer/MintOffer.ts (hunk 2 above)
Comment (maximgeerinck) [reply src/offer/MintOffer.ts:268]: ah, got it now :)

**Comment by maximgeerinck** [comment]
Will try to add some more tests e2e later this weekend

**Review by sneljo1 — APPROVED** [review]

**Pull request merged**